cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.store;
+
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetManagementException;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.NamespaceMeta;
+import co.cask.tephra.TransactionAware;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation for {@link NamespaceStore}.
+ */
+public class DefaultNamespaceStore implements NamespaceStore {
+
+  private static final Id.DatasetInstance APP_META_INSTANCE_ID =
+    Id.DatasetInstance.from(Id.Namespace.SYSTEM, Constants.AppMetaStore.TABLE);
+
+  private final Supplier<NamespaceMetadataStore> apps;
+  private final Supplier<TransactionExecutor> appsTx;
+  private final DatasetFramework dsFramework;
+
+  @Inject
+  public DefaultNamespaceStore(
+    final TransactionExecutorFactory txExecutorFactory,
+    DatasetFramework framework) {
+    this.dsFramework = framework;
+    this.apps = Suppliers.memoize(
+      new Supplier<NamespaceMetadataStore>() {
+        @Override
+        public NamespaceMetadataStore get() {
+          try {
+            Table mdsTable = DatasetsUtil.getOrCreateDataset(
+              dsFramework, APP_META_INSTANCE_ID, "table",
+              DatasetProperties.EMPTY, DatasetDefinition.NO_ARGUMENTS, null);
+            return new NamespaceMetadataStore(mdsTable);
+          } catch (DatasetManagementException | IOException e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      }
+    );
+    this.appsTx = new Supplier<TransactionExecutor>() {
+      @Override
+      public TransactionExecutor get() {
+        return txExecutorFactory.createExecutor(ImmutableList.of((TransactionAware) apps.get()));
+      }
+    };
+  }
+
+  @Override
+  @Nullable
+  public NamespaceMeta create(final NamespaceMeta metadata) {
+    Preconditions.checkArgument(metadata != null, "Namespace metadata cannot be null.");
+    return appsTx.get().executeUnchecked(
+      new TransactionExecutor.Function<NamespaceMetadataStore, NamespaceMeta>() {
```

Comment:
could this use a ``TransactionExecutor.Procedure`` so it doesn't have to return null?